whylog/teacher/rule_validation_problems.py

```diff
@@ -1,110 +1,89 @@
 import itertools
-from abc import ABCMeta, abstractmethod
-from collections import namedtuple
 
 import six
 
-ValidationResult = namedtuple('ValidationResult', ['errors', 'warnings'])
-
 
 class ValidationResult(object):
-    def __init__(self, errors, warnings):
+    def __init__(self, rule_problems, parser_problems, constraint_problems):
         """
-        :param errors: problems preventing form rule saving
-        :param warnings: problems accepted while rule saving
-        :type errors: [RuleValidationProblem]
-        :type warnings: [RuleValidationProblem]
+        :param rule_problems: Problems related to whole rule
+        :param parser_problems: Problems related to parsers, dict[parser id, problem]
+        :param constraint_problems: Problems related to constraints, dict[constraint id, problem]
+        :type rule_problems: list[RuleValidationProblem]
+        :type parser_problems: dict[int, ParserValidationProblem]
+        :type constraint_problems: dict[int, ConstraintValidationProblem]
         """
-        self.errors = errors
-        self.warnings = warnings
+        self.rule_problems = rule_problems
+        self.parser_problems = parser_problems
+        self.constraint_problems = constraint_problems
 
-    @classmethod
-    def result_from_results(cls, results):
-        errors = sum([result.errors for result in results], [])
-        warnings = sum([result.warnings for result in results], [])
-        return ValidationResult(errors, warnings)
+    def is_acceptable(self):
+        parser_problems_list = itertools.chain.from_iterable(six.itervalues(self.parser_problems))
+        constraint_problems_list = itertools.chain.from_iterable(
+            six.itervalues(self.constraint_problems)
+        )  # yapf: disable
 
-    def acceptable(self):
-        return not self.errors
+        all_problems = itertools.chain.from_iterable(
+            [self.rule_problems, parser_problems_list, constraint_problems_list]
```

Comment:
implementation of list may pre-allocate memory for additional items as it is common to append to a list. Tuple implementation will do no such thing. (I am talking about python, not cpython)